A replacement for manage_db.sh + rewrite of migration system documentation

jdavcs opened the pull request:
This is a replacement for the `manage_db.sh` script + new migration documentation.

**Rationale**
The pre-22.05 `manage_db.sh` was a thin wrapper around the SQLAlchemy Migrate CLI runner. With the move to Alembic, this script was rewritten to be an adapter between the old interface supporting 2 essential commands (upgrade/downgrade) and the Alembic CLI. A new script, `run_alembic.sh` was added as a wrapper around Alembic's CLI, to provide access to the full scope of commands provided by Alembic.

However, the `manage_db.sh` script was very limited in what it could do, whereas `run_alembic.sh` required more verbose command syntax, familiarity with advanced Alembic features + an understanding of how Galaxy's data model was implemented (the separation of galaxy and tool shed install models, implemented with Alembic branches).

There was much discussion about designing a better solution that might have these characteristics:
- tool-agnostic interface (not dependent on Alembic or SQLAlchemy Migrate)
- hide the complexity of Alembic branches and the setup of  Galaxy's data model
- provide a set of commands with simple and consistent syntax that enable commonly used functionality. 

This PR proposes a solution.

**Implementation notes**
The script is implemented with argparse (so it provides proper command line help). It supports a set of Alembic CLI commands and command options that may be useful in the context of developing or administering Galaxy. 

The tool targets only the *gxy* data model/migration branch (with a few exceptions, such as running `./db.sh upgrade` without specifying a revision, which is explained in the documentation).

The documentation has been completely rewritten. It provides a better overview of the migration system, detailed descriptions of available commands and options, Galaxy-specific usage examples, etc.

[EDIT: reverted to manage_db.sh after code review.]
~~Why rename `manage_db.sh` to `db.sh`?~~ The new script comes with new commands: a new name makes it explicit, whereas the old name might cause confusion (what version of `manage_db.sh` am I using?). Also, I think, `db.sh` is more succinct: `db.sh upgrade` and `db.sh revision` reads as "database upgrade" and "database revision", whereas `manage_db sh [operation]` is a little less obvious: "manage database upgrade". But that's completely subjective, and I have no objections to keeping `manage_db.sh` if that's a better approach.

The old `manage_db.sh` script translated command line arguments into a format required by the Alembic CLI runner. This new script does not invoke the Alembic script; instead, it invokes Alembic commands programmatically via the Alembic API. As a result, a lot of our code became unnecessary - so it has been deleted. Also, I think the new implementation is more straightforward and less brittle.

**Tests**
Overall approach:
1. Use a blank test database, store revision scripts in a temporary directory, leave the rest unchanged.
2. Use Alembic API for test setup and for accessing the database.
3. Run the `db.sh` command as a subprocess, verify captured output + database state.

In other words, there's almost no additional infrastructure involved: we use Alembic for any setup, and Python's subprocess module for executing the script. 

The tests run against SQLite only. We can run them against Postgres as well, but, I think, that's not necessary: the tests for this script verify that Alembic commands are invoked correctly and produce the expected results; we don't need to test how Alembic interacts with another database.

Here're the details of setting up the staging environment:
- Create staging location (`/tmp`)
- Create test database (SQLite in `/tmp`)
- Copy Galaxy's `alembic.ini` to staging location, overwriting:
  - `sqlalchemy.url` (url of test database)
  - `version_locations` (staging location)
  - `script_location` (point to `lib/galaxy/model/migrations/alembic/` - which is where the real `env.py` is located)
- Create *gxy* and *tsi* branches

**Next steps (separate PRs):**
- [ ] Handle cases when database (one or both) is in pre-22.05 state (container-based solution) + update documentation
- [ ] Same as above, for downgrading to pre-22.05 state + update documentation
- [ ] Update relevant tutorials and slides on GTN

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. Run `./db.sh -h` to view the help and/or `./db.sh [command] -h` to view help for individual commands.
  2. Try running each command. You may use the `-c` option to specify an alternate galaxy.yml file not to modify your database(s).

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

jdavcs committed:
Refactor get_alembic_cfg: make available to other code

jdavcs committed:
Add initial draft

jdavcs committed:
Add logging to script

jdavcs committed:
Add inline testing documentation

jdavcs committed:
Test against 1 or 2 databases

jdavcs committed:
Set dburl in alembic config

jdavcs committed:
Add init command

jdavcs committed:
Fix linting

jdavcs committed:
Fix packages tests: script is one level up

jdavcs committed:
Fix mypy by asserting object before accessing its attributes

jdavcs committed:
Change parsing of --galaxy-config option

Rationale: alternate config file applies to all subcommands, so, I
think, it is more correct to run this:

> ./db.sh -c mygalaxy.yml upgrade some-revision-id

than this:

> ./db.sh upgrade some-revision-id -c mygalaxy.yml

That's also consistent with how alembic handles this option.

jdavcs committed:
Revise db.sh cli command aliases

1. Remove aliases from subcommands that change things (upgrade, downgrade,
revision, init) to reduce risk of accidental usage.

2. Add aliases to readonly subcommands (history, show, version,
   dbversion)

3. Minor edits

jdavcs committed:
Add missing tests

jdavcs committed:
Add --raiseerr option; hide error traceback by default

jdavcs committed:
Misc edits to db.sh

jdavcs committed:
Rename manage_db.sh to manage_toolshed_db.sh. Limit to TS operations.

Becuase everything else is handled via db.sh or run_alembic.sh

jdavcs committed:
Drop create_db.sh, create_db.py (see note)

For ToolShed, we have create_toolshed_db.sh.

For galaxy and toolshed install databases (gxy and tsi branches under
alembic), we use the `init` subcommand of the `db.sh` script (`./db.sh
init`), which has exactly the same functionality as these deleted
scripts.

To the best of my knowledge, these are not used anywhere else (unlike,
for example, scripts/manage_db.py, which is called from the
ansible-galaxy role).

jdavcs committed:
Rename migrate_db.py > run_alembic.py

This is consistent with other similar pairs of files:
[script-name].sh invoking scripts/[script-name].py (check_model, create_roolshed_db, db)

Also, the python file does not require exec permissions.

jdavcs committed:
Remove manage_db_adapter.py

We no longer need to translate the arguments of manage_db.sh into
arguments for alembic. The new facade script for alembic (db.sh)
accesses the alembic api programmatically.

jdavcs committed:
Drop LegacyScripts: no longer used

jdavcs committed:
Remove duplicate migration code

jdavcs committed:
Add type hints

jdavcs committed:
Update test comments

jdavcs committed:
Override script name to display correct usage in cli help

jdavcs committed:
Rewrite migrations documentation

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
 ## Administering Galaxy: upgrading and downgrading the database
 
-To create a database or initialize an empty database, simply start Galaxy. You may also use the `create_db.sh script`. For usage and options, see the inline documentation in that file. Note that this will create and/or initialize new databases (or one combined database) for the `gxy` and `tsi` models. This script does not handle the Tool Shed database, for which you need to use the `create_toolsh ...
+To initialize an empty database (or create a new SQLite database), simply start Galaxy. To upgrade or downgrade an existing database, you'll need to use a script.
```
```suggestion
To initialize an empty database (or create a new SQLite database) start Galaxy. To upgrade or downgrade an existing database, you'll need to use a script.
```

mvdbeek replied:
That's only safe if a single process boots. I'd mention the init command.

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
+To initialize an empty database (or create a new SQLite database), simply start Galaxy. To upgrade or downgrade an existing database, you'll need to use a script.
 
-To upgrade or downgrade an existing database that has been initialized, use the `manage_db.sh` script. For usage and options, see the inline documentation in the file. This script supports a subset of command line options offered previously by SQLAlchemy Migrate. It takes care of adjusting the path, initializing the Python virtual environment, retrieving necessary configuration values, and invoki ...
+To manage database schema migrations, as well as to perform common migration-related operations,
```
```suggestion
```

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
 
-### Examples of usage:
+The `run_alembic.sh` script is a thin wrapper around the Alembic CLI runner. It offers more
+flexibility and the full scope of Alembic's CLI commands and options; however, it requires more detailed command arguments, as well as basic familiarity with [Alembic branches](https://alembic.sqlalchemy.org/en/latest/branches.html). 
```
I think we'd ideally remove the script or place it somewhere that makes it very clear that you don't need to use the script if you're not a developer.

jdavcs replied:
We can move it to the `scripts` directory. I'll make the change. I'll also edit the inline comments + documentation in the readme.

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
-# Galaxy's Migrations System
+# Galaxy Database Schema Migrations
```
Who is the target audience for this document? It seems this is part admin part dev, which isn't pleasant reading for either.

jdavcs replied:
The target audience is devs, maybe admins. I'll move this to a section for developers (or "advanced", or "internals", or something of that sort). It doesn't have to be pleasant reading - it must be helpful reading. It gives the minimum context that should help understand how it works, why there are 2 versions (one in the database, one in the codebase), and why they are expected to be the same. I've had this discussion on many occasions (explaining how it works, etc.), I think it will help to have this as part of the docs (although moving it to a separate section seems to be the right way to do it)

mvdbeek replied:
ok, this looks much better, can you please link it from the docs ?

jdavcs replied:
Will do!

mvdbeek commented on lib/galaxy/model/migrations/util.py:
```diff
@@ -10,24 +10,27 @@
+        "This script is being executed in offline mode and cannot connect to the database. "
+        f"Therefore, `{code}` returns `{return_value}` by default."
+    )
+    log.debug(msg)
```
```suggestion
    log.warning(msg)
```

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
 
-`sh manage_db.sh downgrade --version=xyz789 -c mygalaxy.yml install`
+This script offers a set of common database schema migration operations that are executed on the *gxy* branch, which, in the vast majority of cases, is all you need to develop and administer Galaxy.
+To run operations on the *tsi* branch, you need to use `run_alembic.sh`.
```
Can't we just figure out what needs to be done given the Galaxy configuration ? so `manage_db.sh upgrade` upgrades everything configured for Galaxy at this point, either the combined models or the two separate databases. Or maybe I don't understand what 

> To run operations on the *tsi* branch

means ?

jdavcs replied:
That's exactly what it does: `db.sh upgrade` handles both, connecting to one db or two dbs, depending on galaxy's configuration. Only devs would need to touch the tsi branch (if ever). "Run operations on the tsi branch" refers to running migration operations on the tool shed install model without touching the galaxy model. I'll simplify this section.

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
+    history (h)         List revision scripts in chronological order
+    show (s)            Show the revision(s) denoted by the given symbol
+    revision            Create a new revision file
+    init                Initialize empty database(s) for both branches
```
```suggestion
    init                Initialize empty database(s)
```

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
+    show (s)            Show the revision(s) denoted by the given symbol
+    revision            Create a new revision file
+    init                Initialize empty database(s) for both branches
+                        (create database objects for gxy and tsi branch)
```
```suggestion
```

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
+    dbversion (dv)      Show the current revision for Galaxy's database
+    history (h)         List revision scripts in chronological order
+    show (s)            Show the revision(s) denoted by the given symbol
+    revision            Create a new revision file
```
```suggestion
```

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
+    version (v)         Show the head revision in the migrations script directory
+    dbversion (dv)      Show the current revision for Galaxy's database
+    history (h)         List revision scripts in chronological order
+    show (s)            Show the revision(s) denoted by the given symbol
```
```suggestion
```

mvdbeek replied:
I don't think these and the options below should be shown to admins, that's not something they can act on. If you want to keep these in the script maybe prefix this with a big warning flag that this is for developers ?

jdavcs replied:
I haven't considered separating the documentation (and available commands) into docs/commands for admins, and for devs. As I've mentioned in a previous comment, I agree that makes sense. Let me think about it and I'll try to come up with a way to do it.

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
 
-Upgrade gxy to head revision:
+Some of the commands accept revision identifiers as arguments. A revision is usually identified by a
+12-digit hexadecimal number (i.e., `6a67bf27e6a6`). Anytime you need to refer to a specific
```
I'd skip the partial number stuff, that's a nice thing but if you're reading the document I don't think that's relevant.

jdavcs replied:
Hmm.. I think it's helpful, especially for devs (given that we are now dealing with 12-digit hex numbers). But again, I'll move the description into a different section; the functionality will be there - it's how alembic processes revision identifiers.

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
+#### Relative migration identifiers
 
-Upgrade branch to a specific revision:
+You may also use Alembic's syntax for relative migration identifiers for the upgrade/downgrade commands:
```
Again, a nice shortcut for devs, but I don't think an admin would ever want to do this. They may think `-1/+1` corresponds to the previous/next release instead of revision.

jdavcs replied:
Yeah, I see your point. I'll edit this. [same as my previous comment]

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
+
+#### Upgrading
+
+Upgrade to the head revision (both, *gxy* and *tsi* branches):
+
+`./run_alembic.sh upgrade heads`
+
+Upgrade to the head revision (*gxy* branch):
+
+`./run_alembic.sh upgrade gxy@head`
+
+Upgrade to 1 revision above the current (*tsi* branch):
+
+`./run_alembic.sh upgrade tsi@+1`
+
+Upgrade to a specific revision:
+
+`./run_alembic.sh upgrade [revision identifier]`
+
+Upgrade to 1 revision above a specific revision:
+
+`./run_alembic.sh upgrade [revision identifier]+1`
```
I'd only list the stuff admins may want to do

jdavcs replied:
Will separate. [same as my previous comment]

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
+`./run_alembic.sh downgrade [revision identifier]-1 `
 
-When creating new revisions, as with upgrading/downgrading, you need to indicate the target branch. Use the `run_alembic.sh` script.
+#### Creating new revisions
```
I'd split this document into a short admin-centric version and a version for developers.

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
+
+   Once your database has the latest SQLAlchemy Migrate version, switch back to your current branch (22.05 or more recent).
+
+5. Run `db.sh upgrade`.
```
Maybe keep the script at manage_db.sh

mvdbeek commented on lib/galaxy/model/migrations/README.md:
```diff
@@ -1,115 +1,391 @@
+
+5. Run `db.sh upgrade`.
+
+## Developing Galaxy: creating new revisions
```
Seems redundant with `Creating new revisions` .

jdavcs replied:
I don't think so: one describes the revision command and its options, the other one describes the process (i.e., the command + what file is generated and where it's put by alembic + recommendation to use our abstractions + tests + what to run next, etc. But again, maybe separating the docs will make this appear less redundant.

mvdbeek commented on lib/galaxy/model/migrations/scripts.py:
```diff
@@ -319,6 +212,19 @@ def run_upgrade(self, gxy_db_url=None, tsi_db_url=None):
 
+    def rename_config_argument(self, argv: List[str]) -> None:
+        """
+        Rename the optional config argument: we can't use '-c' because that option is used by Alembic.
```
Is that still needed given that you're using the API now ?

jdavcs replied:
It could be, I'll check. But this might be used by manage_db.py, which is invoked used by the ansible role.

jdavcs replied:
Yep, it's used by the ansible role.

jdavcs committed:
Update lib/galaxy/model/migrations/README.md

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

jdavcs committed:
Update lib/galaxy/model/migrations/README.md

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

jdavcs committed:
Update lib/galaxy/model/migrations/README.md

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

jdavcs committed:
Update lib/galaxy/model/migrations/README.md

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

jdavcs committed:
Rename db.sh > manage_db.sh, following code review feedback

jdavcs committed:
Remove run_alembic.sh from root; user scripts/run_alembic.py instead.

Rationale: move developer-focused scripts out of galaxy root.

jdavcs committed:
Split script into admin and dev

TODO: remove duplication between db.py and db_dev.py (split into
multiple commits is intentional)

jdavcs committed:
Remove duplication

jdavcs committed:
Split documentation into admin and dev sections

jdavcs committed:
Point to init command as safer way to initialize empty database
